airflow/datasets/__init__.py

```diff
@@ -106,8 +112,90 @@ class DatasetAny(_DatasetBooleanCondition):
 
     agg_func = any
 
+    def __init__(self, *objects: Dataset | DatasetAny | DatasetAll) -> None:
+        """Initialize with one or more Dataset, DatasetAny, or DatasetAll instances."""
+        super().__init__(*objects)
+
+    def __or__(self, other):
```

Comment:
```suggestion
    def __or__(self, other: Dataset | DatasetAny | DatasetAll) -> DatasetAny:
```

Reply:
or perhaps `BaseDatasetEventInput` would work?

Reply:
Yes that’s why I added that base protocol

Reply:
Changed it to `BaseDatasetEventInput`